rfcs/2020-11-02-remap-infallible-functions.md

```diff
@@ -0,0 +1,371 @@
+unexpected situations.
+
+If whatever value the `message` field contains cannot be parsed as valid JSON,
+the `parse_json` function returns an error.
+
+This results in an outcome where the original event is left untouched, the
+`message` field still contains the original value, and no `timestamp` field
+exists (assuming none existed to start with).
+
+A use-case such as the above is quite common. Fields often need to be parsed as
+JSON values, but at the same time those fields are highly dynamic and might not
+always be considered valid JSON, for example if a bug in an application starts
+generating invalid stringified JSON output.
+
+Users need a way to handle such situations, without having to resort to complex
+solutions.
+
+### Infallible Function Variants
+
+The solution proposed by this RFC is to introduce the concept of _infallible
+function variants_.
+
+An infallible function variant is an exact copy of an existing fallible
+function, but it is guaranteed to never fails.
+
+The infallible function must be as easy (or easier) to call than the fallible
+function, without any significant syntactic complexity.
+
+### To Bang Or Not To Bang
+
+The proposal is to introduce the concept of the bang (`!`) function identifier
+additive to determine as a user if a function should be allowed to fail or not.
+
+Function identifiers with a bang are considered to be fallible, while functions
+without are not:
+
+```javascript
+.message = parse_json(.message) // cannot fail
+.counter = 1 + to_int!(.counter) // allowed to fail
+.timestamp = now()
+```
+
+In the above example, `parse_json` is written without a bang, and will thus
+never fail, whereas `to_int!` includes the bang, and will fail the program (and
+leave the event unchanged) if `.counter` cannot be coerced into an integer.
+
+### Infallible Function Result
+
+So far so good, but telling a program you don't want JSON parsing the fail does
+not magically make any input valid JSON.
+
+Given the original example, any badly formatted input values are still
+considered invalid JSON and the parser is thus unable to return a representative
+JSON value.
+
+So what should the function return instead?
+
+Infallible functions are allowed to return `Option<Value>`. In other words, they
+can return "nothing", if there isn't anything useful to return.
+
+### How To Handle "Nothing"
+
+Let's take a look at the original example again:
+
+```javascript
+.message = parse_json(.message)
+.timestamp = now()
+```
+
+If `message` contains any data that `parse_json` can't handle, it will return
+`None`.
+
+How that `None` gets processed is up to the expression that processes the value.
+
+For example, in the case of the `Assignment` expression (which is what's called
+when doing `.message = …`), `None` values skip assigning the value to the target
+path, returning `None` itself.
+
+Given this, the outcome of the above example would change from "exception,
```

Comment:
I actually find it confusing that `.message = parse_json(.message)` would not assign `.message` if parsing fails. I would expect it to receive the value of `null`. That is, I would not expect the assignment to be conditional on whether the function executes successfully.

Reply:
I agree. I think for the exact reason mentioned on line 168 that it should pass the `undefined` value on. (Note `null` is a valid value, so we do need the distinction between `null` and `undefined` here).

Reply:
I think we have to weigh this confusion vs adding a second meaning of `null` as "operation failed". If we don't differentiate between the two, operations that legitimately return `null` could trigger unexpected behavior.